src/commands/force/source/delete.ts

```diff
@@ -0,0 +1,206 @@
+  public static readonly requiresProject = true;
+  public static readonly requiresUsername = true;
+  public static readonly flagsConfig: FlagsConfig = {
+    checkonly: flags.boolean({
+      char: 'c',
+      description: messages.getMessage('flags.checkonly'),
+      longDescription: messages.getMessage('flags.checkonlyLong'),
+    }),
+    wait: flags.minutes({
+      char: 'w',
+      default: Duration.minutes(Delete.DEFAULT_SRC_WAIT_MINUTES),
+      min: Duration.minutes(1),
+      description: messages.getMessage('flags.wait'),
+    }),
+    testlevel: flags.enum({
+      char: 'l',
+      description: messages.getMessage('flags.testLevel'),
+      options: ['NoTestRun', 'RunLocalTests', 'RunAllTestsInOrg'],
+      default: 'NoTestRun',
+    }),
+    noprompt: flags.boolean({
+      char: 'r',
+      description: messages.getMessage('flags.noprompt'),
+    }),
+    metadata: flags.array({
+      char: 'm',
+      description: messages.getMessage('flags.metadata'),
+      exclusive: ['manifest', 'sourcepath'],
+    }),
+    sourcepath: flags.array({
+      char: 'p',
+      description: messages.getMessage('flags.sourcepath'),
+      longDescription: messages.getMessage('flags.sourcepathLong'),
+      exclusive: ['manifest', 'metadata'],
+    }),
+    verbose: flags.builtin({
+      description: messages.getMessage('flags.verbose'),
+    }),
+  };
+  protected xorFlags = ['metadata', 'sourcepath'];
+  protected readonly lifecycleEventNames = ['predeploy', 'postdeploy'];
+  private sourceComponents: SourceComponent[];
+  private isRest = false;
+
+  private updateDeployId = once((id) => {
+    this.displayDeployId(id);
+    this.setStash(id);
+  });
+
+  public async run(): Promise<DeployCommandResult> {
+    await this.delete();
+    this.resolveSuccess();
+    return this.formatResult();
+  }
+
+  protected async delete(): Promise<void> {
+    // verify that the user defined one of: metadata, sourcepath
+    this.validateFlags();
+
+    this.componentSet = await ComponentSetBuilder.build({
+      apiversion: this.getFlag<string>('apiversion'),
+      sourceapiversion: await this.getSourceApiVersion(),
+      sourcepath: this.getFlag<string[]>('sourcepath'),
+      metadata: this.flags.metadata && {
+        metadataEntries: this.getFlag<string[]>('metadata'),
+        directoryPaths: this.getPackageDirs(),
+      },
+    });
+
+    this.sourceComponents = this.componentSet.getSourceComponents().toArray();
+
+    if (!this.sourceComponents.length) {
+      // if we didn't find any components to delete, let the user know and exit
+      // matches toolbelt
+      this.ux.styledHeader(chalk.blue('Deleted Source'));
+      this.ux.log('No results found');
+      this.exit(0);
+    }
+
+    // create a new ComponentSet and mark everything for deletion
+    const cs = new ComponentSet([]);
+    this.sourceComponents.map((component) => {
+      cs.add(component, true);
+    });
+    this.componentSet = cs;
```

Comment:
We can't do it this way because we'll lose `apiVersion` and `sourceApiVersion` by creating a new `ComponentSet`.  We should modify `ComponentSetBuilder` to take another optional option to mark them for delete.

Reply:
[destructive change files](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_deploy_deleting_files.htm) don't include an api version in them, it's recommend to use an empty package, with the api, if necessary

Reply:
Ok but I still think creating multiple ComponentSets is an anti-pattern.  If we don't think other commands will need ComponentSetBuilder modified for deletes then this is ok.